### PR TITLE
PP-9785 Do not throw exceptions for Worldpay order not found query response

### PR DESCRIPTION
**Pull request by kbottla**
## WHAT

- Worldpay connection timed out payments end up in authorisation error states and the requests may never have reached Worldpay. When various tasks (gateway cleanup sweep, discrepancy checker) query Worldpay for the transaction, an error response is returned with code `5` and the message `Could not find payment for order`.
- Currently, WebApplicationException is thrown by the Worldpay query payment method and is causing 500 errors for the discrepancy checker. Gateway cleanup service catches exceptions and continues onto the next payment (without updating charge status). So auth errored charges (for which is there is no Worldpay order) are being retried continuously and new charges are not being cleaned up (as the gateway clean service is attempting to query the same charges for every invoke and also has a limit of 100 charges)
- So returning `ChargeQueryResponse` when Worldpay returns `Could not find payment for order` error.
    - The discrepancy checker will return the message and is displayed in Toolbox
    - Gateway cleanup service will update charge status to `AUTHORISATION_ERROR_CHARGE_MISSING` (which will be expunged by connector)

### CODE REVIEW

**Comment by alexbishop1** [comment]
What happens when Worldpay’s order inquiry endpoint is down (which happens quite often)? In those cases retrying again later can actually be useful.

**Comment by kbottla** [comment]
@alexbishop1 I am expecting a different error message than what is handled here https://github.com/alphagov/pay-connector/blob/60b70545a83703c7826b983154081e4bce59fd52/src/main/java/uk/gov/pay/connector/gateway/worldpay/WorldpayPaymentProvider.java#L155 when Worldpay inquiry endpoint is down. 

It would be ideal to have a retrying mechanism and ignore any charges tried n number of times. But that is a bigger piece of work we can think of separately.

**Comment by alexbishop1** [comment]
Yeah, I agree it’s a bigger piece of work. I can’t remember what the standard failure mode is for when the order inquiry endpoint is down — does it do something different to claiming it can’t find the order?

**Comment by kbottla** [comment]
I see `error code: 8, error: Order inquiries have been disabled for all merchants` in the logs for the maintenance window  `12 August 2021 18:00 (UTC) - 21:00 (UTC)`. 

This lines up with Worldpay docs https://developer.worldpay.com/docs/wpg/troubleshoot

Error code no. | Error type | Comments
-- | -- | --
8 | Service not available. | The service is currently unavailable.

**Comment by alexbishop1** [comment]
OK, cool, so this is sufficiently narrowly-targeted.

File: src/main/java/uk/gov/pay/connector/gateway/worldpay/WorldpayPaymentProvider.java [hunk 1]
```diff
@@ -148,6 +148,18 @@ public ChargeQueryResponse queryPaymentStatus(ChargeQueryGatewayRequest chargeQu
 
                     return new ChargeQueryResponse(mappedStatus, worldpayQueryResponse);
                 })
+                .or(() -> {
+                            return worldpayGatewayResponse.getGatewayError()
+                                    .map(gatewayError -> {
+                                                if (gatewayError.getMessage() != null
+                                                        && gatewayError.getMessage().contains("Could not find payment for order")) {
```
Comment (alexbishop1) [inline src/main/java/uk/gov/pay/connector/gateway/worldpay/WorldpayPaymentProvider.java]: Can we…

- … also check the code is 5?
- … extract the error message into a constant and put a comment linking to some appropriate Worldpay documentation (it seems https://developer.worldpay.com/docs/wpg/manage/inquiryrequests is the best available — it describes the string under the ‘Getting an error?’ expand-o-matic)?

File: src/test/java/uk/gov/pay/connector/gateway/worldpay/WorldpayPaymentProviderTest.java [hunk 2]
```diff
@@ -580,14 +582,53 @@ void should_successfully_query_payment_status() throws Exception {
...
+        assertThat(chargeQueryResponse.foundCharge(), is(false));
+        assertThat(chargeQueryResponse.getGatewayError().isPresent(), is(true));
+        assertThat(chargeQueryResponse.getGatewayError().get().getMessage(), is("Worldpay query response (error code: 5, error: Could not find payment for order)"));
+    }
+
+    @Test
+    void query_payment_status_should_throw_error_when_worldpay_returns_error_with_unknown_message() throws Exception {
+        when(response.getEntity()).thenReturn(
+                load(WORLDPAY_ERROR_RESPONSE)
+                        .replace("{{errorDescription}}", "some error message")
```
Comment (alexbishop1) [inline src/test/java/uk/gov/pay/connector/gateway/worldpay/WorldpayPaymentProviderTest.java]: We could maybe use the specific message we expect when the order inquiry endpoint is down.

**Review by alexbishop1 — CHANGES_REQUESTED** [review]
Looks fine. Just a couple of things that may help our future selves.

File: src/test/java/uk/gov/pay/connector/gateway/worldpay/WorldpayPaymentProviderTest.java (hunk 2 above)
Comment (kbottla) [reply src/test/java/uk/gov/pay/connector/gateway/worldpay/WorldpayPaymentProviderTest.java]: Updated in https://github.com/alphagov/pay-connector/compare/60b70545a83703c7826b983154081e4bce59fd52..063f8416eda0e04efa6f707f103933d8b8de273f

File: src/main/java/uk/gov/pay/connector/gateway/worldpay/WorldpayPaymentProvider.java (hunk 1 above)
Comment (kbottla) [reply src/main/java/uk/gov/pay/connector/gateway/worldpay/WorldpayPaymentProvider.java]: Updated https://github.com/alphagov/pay-connector/compare/60b70545a83703c7826b983154081e4bce59fd52..063f8416eda0e04efa6f707f103933d8b8de273f

**Review by alexbishop1 — APPROVED** [review]

**Pull request merged**